resolver/resolver.go

```diff
@@ -354,111 +347,117 @@ func (res *Resolver) HandleNonMesos(w dns.ResponseWriter, r *dns.Msg) {
-		for _, a := range rs.As[dom] {
-			rr, err := res.formatA(dom, a)
-			if err != nil {
-				logging.Error.Println(err)
-			} else {
-				m.Answer = append(m.Answer, rr)
-			}
+	reply(w, m)
+}
 
+func (res *Resolver) handleSRV(rs *records.RecordGenerator, name string, m, r *dns.Msg) {
+	for _, srv := range rs.SRVs[name] {
+		srvRR, err := res.formatSRV(r.Question[0].Name, srv)
+		if err != nil {
+			logging.CurLog.MesosFailed.Inc()
+			logging.Error.Println(err)
+			continue
 		}
-	}
-
-	// SOA requests
-	if (qType == dns.TypeSOA) || (qType == dns.TypeANY) {
-		rr, err := res.formatSOA(r.Question[0].Name)
+		m.Answer = append(m.Answer, srvRR)
+		// return one corresponding A record add additional info
+		host := strings.Split(srv, ":")[0]
+		if len(rs.As[host]) == 0 {
+			continue
+		}
+		aRR, err := res.formatA(host, rs.As[host][0])
 		if err != nil {
+			logging.CurLog.MesosFailed.Inc()
 			logging.Error.Println(err)
-		} else {
-			m.Ns = append(m.Ns, rr)
+			continue
 		}
+		m.Extra = append(m.Extra, aRR)
 	}
+}
 
-	// NS requests
-	if (qType == dns.TypeNS) || (qType == dns.TypeANY) {
-		rr, err := res.formatNS(r.Question[0].Name)
-		logging.VeryVerbose.Println("NS request")
+func (res *Resolver) handleA(rs *records.RecordGenerator, name string, m *dns.Msg) {
+	for _, a := range rs.As[name] {
+		rr, err := res.formatA(name, a)
 		if err != nil {
+			logging.CurLog.MesosFailed.Inc()
 			logging.Error.Println(err)
-		} else {
-			m.Ns = append(m.Ns, rr)
+			continue
 		}
+		m.Answer = append(m.Answer, rr)
 	}
+}
 
-	// shuffle answers
-	m.Answer = shuffleAnswers(m.Answer)
-	// tracing info
-	logging.CurLog.MesosRequests.Inc()
+func (res *Resolver) handleSOA(m, r *dns.Msg) {
+	rr, err := res.formatSOA(r.Question[0].Name)
+	if err != nil {
+		logging.CurLog.MesosFailed.Inc()
+		logging.Error.Println(err)
+		return
+	}
+	m.Ns = append(m.Ns, rr)
+}
 
+func (res *Resolver) handleNS(m, r *dns.Msg) {
+	rr, err := res.formatNS(r.Question[0].Name)
+	logging.Error.Println("NS request")
 	if err != nil {
 		logging.CurLog.MesosFailed.Inc()
-	} else if (qType == dns.TypeAAAA) && (len(rs.SRVs[dom]) > 0 || len(rs.As[dom]) > 0) {
-		// correct handling of AAAA if there are A or SRV records
-		m.Authoritative = true
-		// set NOERROR
-		m.SetRcode(r, 0)
-		// leave answer empty (NOERROR --> NODATA)
+		logging.Error.Println(err)
+		return
+	}
+	m.Ns = append(m.Ns, rr)
+}
 
-	} else {
-		// no answers but not a {SOA,SRV} request
-		if len(m.Answer) == 0 && (qType != dns.TypeSOA) && (qType != dns.TypeNS) && (qType != dns.TypeSRV) {
-			// set NXDOMAIN
-			m.SetRcode(r, 3)
-
-			rr, err := res.formatSOA(r.Question[0].Name)
-			if err != nil {
-				logging.Error.Println(err)
-			} else {
-				m.Ns = append(m.Ns, rr)
-			}
+func (res *Resolver) handleEmpty(rs *records.RecordGenerator, name string, m, r *dns.Msg) {
+	qType := r.Question[0].Qtype
+	switch qType {
+	case dns.TypeSOA, dns.TypeNS, dns.TypeSRV:
+		return
```

Comment:
These three record types increase the MesosSuccess metric before, but don't do it now anymore. Intentional?